api/src/experimental/trace/SugaredTracer.ts

```diff
@@ -0,0 +1,213 @@
+   *
+   * @param name The name of the span
+   * @param [options] SugaredSpanOptions used for span creation
+   * @param [context] Context to use to extract parent
+   * @param fn function called in the context of the span and receives the newly created span as an argument
+   * @returns Span The newly created span
+   * @example
+   *     const something = tracer.withSpan('op', span => {
+   *      // do some work
+   *     });
+   * @example
+   *     const something = await tracer.withSpan('op', span => {
+   *      // do some async work
+   *     });
+   */
+  withSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    fn: F
+  ): ReturnType<F>;
+  withSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    options: SugaredSpanOptions,
+    fn: F
+  ): ReturnType<F>;
+  withSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    options: SugaredSpanOptions,
+    context: Context,
+    fn: F
+  ): ReturnType<F>;
+  withSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    options: SugaredSpanOptions,
+    context: Context,
+    fn: F
+  ): ReturnType<F>;
+  withSpan<F extends (span: Span) => ReturnType<F>>(
+    name: string,
+    arg2: SugaredSpanOptions | F,
+    arg3?: Context | F,
+    arg4?: F
+  ): ReturnType<F> {
+    const { opts, ctx, fn } = massageParams(arg2, arg3, arg4);
+
+    const span = this._tracer.startSpan(name, opts, ctx);
+    return handleFn(span, opts, fn) as ReturnType<F>;
+  }
+}
+
+/**
+ * Massages parameters of withSpan and withActiveSpan to allow signature overwrites
+ * @param arg
+ * @param arg2
+ * @param arg3
+ */
+function massageParams<F extends (span: Span) => ReturnType<F>>(
+  arg: F | SugaredSpanOptions,
+  arg2?: F | Context,
+  arg3?: F
+) {
+  let opts: SugaredSpanOptions | undefined;
+  let ctx: Context | undefined;
+  let fn: F;
+
+  if (!arg2 && !arg3) {
+    fn = arg as F;
+  } else if (!arg3) {
+    opts = arg as SugaredSpanOptions;
+    fn = arg2 as F;
+  } else {
+    opts = arg as SugaredSpanOptions;
+    ctx = arg2 as Context;
+    fn = arg3 as F;
+  }
+  opts = opts ?? {};
+  ctx = ctx ?? context.active();
+
+  return { opts, ctx, fn };
+}
+
+/**
+ * Executes fn, returns results and runs onException in the case of exception to allow overwriting of error handling
+ * @param span
+ * @param opts
+ * @param fn
+ */
+function handleFn<F extends (span: Span) => ReturnType<F>>(
+  span: Span,
+  opts: SugaredSpanOptions,
+  fn: F
+): ReturnType<F> {
+  const onException = opts.onException ?? defaultOnException;
+
+  try {
+    const ret = fn(span) as Promise<ReturnType<F>>;
+    // if fn is an async function, attach a recordException and spanEnd callback to the promise
+    if (typeof ret?.then === 'function' && typeof ret?.catch === 'function') {
+      return ret
+        .catch((e: Error) => {
+          onException(e, span);
+          throw e;
+        })
+        .finally(() => span.end()) as ReturnType<F>;
```

Comment:
To mitigate the failure, we can instead utilize `.then` only.

```suggestion
    if (typeof ret?.then === 'function') {
      return ret
        .then((val) => {
          span.end();
          return val;
        }, (e: Error) => {
          onException(e, span);
          span.end();
          throw e;
        }) as ReturnType<F>;
```